concurrency/runner.go

```diff
@@ -0,0 +1,108 @@
+package concurrency
+
+import (
+	"context"
+	"sync"
+
+	"golang.org/x/sync/errgroup"
+
+	dskitMath "github.com/grafana/dskit/math"
+	"github.com/grafana/dskit/multierror"
+)
+
+// ForEachUser runs the provided userFunc for each userIDs up to concurrency concurrent workers.
+// In case userFunc returns error, it will continue to process remaining users but returns an
+// error with all errors userFunc has returned.
+func ForEachUser(ctx context.Context, userIDs []string, concurrency int, userFunc func(ctx context.Context, userID string) error) error {
+	if len(userIDs) == 0 {
+		return nil
+	}
+
+	// Push all jobs to a channel.
+	ch := make(chan string, len(userIDs))
+	for _, userID := range userIDs {
+		ch <- userID
+	}
+	close(ch)
+
+	// Keep track of all errors occurred.
+	errs := multierror.MultiError{}
+	errsMx := sync.Mutex{}
+
+	wg := sync.WaitGroup{}
+	for ix := 0; ix < dskitMath.Min(concurrency, len(userIDs)); ix++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+
+			for userID := range ch {
+				// Ensure the context has not been canceled (ie. shutdown has been triggered).
+				if ctx.Err() != nil {
+					break
+				}
+
+				if err := userFunc(ctx, userID); err != nil {
+					errsMx.Lock()
+					errs.Add(err)
+					errsMx.Unlock()
+				}
+			}
+		}()
+	}
+
+	// wait for ongoing workers to finish.
+	wg.Wait()
+
+	if ctx.Err() != nil {
+		return ctx.Err()
+	}
+
+	errsMx.Lock()
+	defer errsMx.Unlock()
```

Comment:
No need for locking here. `wg.Wait()` call (together with using `defer wg.Done()`)  already establishes happens-before relation between goroutines adding errors, and main goroutine calling `errs.Err()`. (`errsMx` is still needed inside goroutines though)